app/uk/gov/hmrc/tai/controllers/income/IncomeController.scala

```diff
@@ -48,17 +53,72 @@ class IncomeController @Inject()(incomeService: IncomeService,
       } recoverWith taxAccountErrorHandler
   }
 
-  def taxCodeIncomesForYear(nino: Nino, year:TaxYear): Action[AnyContent] = authentication.async {
+  def taxCodeIncomesForYear(nino: Nino, year: TaxYear): Action[AnyContent] = authentication.async {
     implicit request =>
       incomeService.taxCodeIncomes(nino, year).map {
         case Seq() => NotFound
         case taxCodeIncomes => Ok(Json.toJson(ApiResponse(Json.toJson(taxCodeIncomes)(Writes.seq(taxCodeIncomeSourceWrites)), Nil)))
-      }recoverWith taxAccountErrorHandler
+      } recoverWith taxAccountErrorHandler
+  }
+
+  def matchedTaxCodeIncomesForYear(nino: Nino, year: TaxYear, incomeType: String, status: String): Action[AnyContent] = authentication.async {
+    implicit request =>
+
+      def filterMatchingEmploymentsToIncomeSource(employments: Seq[Employment], filteredTaxCodeIncomes: Seq[TaxCodeIncome]): Seq[IncomeSource] =
+        filteredTaxCodeIncomes.flatMap { income =>
+          employments.flatMap(emp =>
+            income.employmentId.fold(Seq.empty[IncomeSource]) {
+              id => if (id == emp.sequenceNumber) Seq(IncomeSource(income, emp)) else Seq.empty[IncomeSource]
+            }
+          )
```

Comment:
I don't think fold reads well here as the failing fold is an empty and so is the inner case, a map with a pattern match should do